### PR TITLE
Template Diagnostic Improvements

### LINKED ISSUE #17959: Diagnostics for template keywords showing up inside local classes aren't good
|  |  |
| --- | --- |
| Bugzilla Link | [17585](https://llvm.org/bz17585) |
| Version | unspecified |
| OS | Linux |
| CC | @DougGregor,@zygoloid |

## Extended Description 
consider:
```cpp
void foo() {
  template <typename> int i;
}
```
we issue:

```console
error: expected expression
```

gcc provides a better error:

```console
error: a template declaration cannot appear at block scope
```

another case:

```cpp
void foo() {
  struct S { template <typename> int i; };
}
```
we issue:

```console
error: member 'i' declared as a template
```

gcc provides:

```console
invalid declaration of member template in local class
```

These plain english messages are very nice, we should do something similar.

### PR DESCRIPTION
**Pull request by bradenhelmer**
It turns out `SemaTemplate` handles this type of diagnostic already, however when  template gets encountered, it never gets parsed as a possible statement or declaration, only as an expression. 

Would like some feedback as I am unsure if this is the right fix. Thanks

Fixes #17959.

### CODE REVIEW

**Review by mizvekov — COMMENTED** [review]
Hi.

The change looks good.

There are some other examples of bad diagnostics in that issue, and it would be ideal if they are preserved in case this PR closes that issue.

Please also add an entry to `clang/docs/ReleaseNotes.rst`.

**Comment by bradenhelmer** [comment]
This is handling the other non-local case now, and the release notes have been updated. The local class case was already being handled.

**Review by mizvekov — APPROVED** [review]
LGTM, Thanks!

Please give it a couple of days before merging, in case anyone else wants to chime in.

**Review by mizvekov — APPROVED** [review]
These updated test changes still LGTM.

**Review by cor3ntin — APPROVED** [review]
LGTM.

**Comment by bradenhelmer** [comment]
I don't have write access, if someone could merge that would be great!

**Pull request merged**
